agent/structs/structs.go

```diff
@@ -334,7 +335,10 @@ func (r *DCSpecificRequest) CacheInfo() cache.RequestInfo {
 	// To calculate the cache key we only hash the node filters. The
 	// datacenter is handled by the cache framework. The other fields are
 	// not, but should not be used in any cache types.
-	v, err := hashstructure.Hash(r.NodeMetaFilters, nil)
+	v, err := hashstructure.Hash([]interface{}{
+		r.NodeMetaFilters,
+		r.Filter,
```

Comment:
Think this is sensible and sufficient for now. I wonder if there is any way to normalize the filter string in some way so that silly things like space vs no space or different delimiters or clause ordering that makes no logical difference doesn't result in duplicate cache entries? My guess is that the effort is only worth it if we see it being a problem for people in practice.

Reply:
One bit of follow on work I was thinking of would be to allow a configuration setting to push filtering to the clients instead of servers. In that case none of this would matter as it would just execute the filter on the client against the already cached results.

One other thought was to do the parsing of the filter into the AST on the clients and then shipping the AST off to the servers. This would eliminate the overhead of parsing from the servers and have the added benefit of automatically normalizing the filter (whitespace, unnecessary parens, unnecessary double "not" operators etc can all be elided).

The main downside is that older clients with newer servers could only use the filter language features known to their clients instead of the updated features/syntax known by servers.

I don't think this would be a deal breaker or a good enough reason not to ship a parsed filter to the servers.

Other notes are that shipping the AST rather than the string is going to use a little more network bandwidth for each request. I don't think it should be enough to matter though. Also, it would mean needing to figure out serialization/deserialization for msgpack. Nothing that cannot be done. It would just be a little more work.